### PR TITLE
coqdep: only output vos when passed -vos

### PR DESCRIPTION
**Pull request by SkySkimmer**
This fixes dune. TBH the problem is that dune is too strict, but we
can't go back in time to change it.

### CODE REVIEW

**Comment by ejgallego** [comment]
This is in principle fine but we need to wait for the other PR to be approved too.

**Comment by SkySkimmer** [comment]
>This is in principle fine but we need to wait for the other PR to be approved too.

I don't see why, the 2 PRs don't really interact.

**Comment by ejgallego** [comment]
> I don't see why, the 2 PRs don't really interact.

Indeed, but a rebase of one of them maybe necessary; anyways I'm taking care of the `.vos` patches now.

**Review by ejgallego — APPROVED** [review]
Thanks for taking care of this; I think the fix is good for now.

We must also fix https://github.com/ocaml/dune/issues/2842 by making Dune's parser more resistant.

**Comment by ejgallego** [comment]
Re-running vst as it is not clear the failure is independent from this PR.

**Review by gares — APPROVED** [review]

**Comment by charguer** [comment]
Hum, may I intervene here, even though no one though it would be a good idea to notify me of the issue, which I discovered after deep investigations to figure out why I wouldn't get the proper dependencies generated. 

The issue now is that if the user calls `coqdep` without `-vos` then all the vos stuff becomes completely bogus, as dependencies for `.vos` files are missing. At the very least, it deserves an update to the documentation of both `coqdep` and the section on `vos` files, for people who use `coqdep` directly and not `coq_makefile`.

**Comment by charguer** [comment]
Follow up question: do you plan for the `-vos` flag of `coqdep` to be something that remains there forever? Or will it ultimately disappear? 

Also, one thing that's not clear to me is why you have no issues with dependencies on `.vio` files, (why is it not an issue for `dune` whereas `vos` dependencies are?)

**Comment by SkySkimmer** [comment]
>Follow up question: do you plan for the -vos flag of coqdep to be something that remains there forever? Or will it ultimately disappear?

Once dune can handle the extra lines we can remove it.

>Also, one thing that's not clear to me is why you have no issues with dependencies on .vio files, (why is it not an issue for dune whereas vos dependencies are?)

Because dune specifically expects coqdep to output 1 or 2 lines
https://github.com/ocaml/dune/blob/ea54f43a2ccf7d96d6ce91b7126f8b867be4c467/src/dune/coq_rules.ml#L42

**Comment by charguer** [comment]
Thanks for the precise pointer. I find it hard to believe that somebody wrote such a strict code without anticipating for any potential change in the behavior of `coqdep`. Or perhaps the author thought: "well, the day it changes, we'll adapt our code". Yet, if I understand well the issue at play, this line of reasoning fails to take the backward compatibility issue in mind. Maybe it should be better advertised to tools that depend on Coq of this kind of caveat.

**Comment by ejgallego** [comment]
The code is strict on purpose; it is better to fail hard than to have your tooling picking up unexpected inputs.

IMO an experimental patch such as the `vos` code should produce no observable change unless it is enabled explicitly by the user. Unfortunately this is not the case but we should get as close as possible.

**Comment by silene** [comment]
"Strict" is a bit of an overstatement, since the code does not even check that what it got are the dependencies of the `.vo` file.

**Comment by ejgallego** [comment]
> "Strict" is a bit of an overstatement, since the code does not even check that what it got are the dependencies of the `.vo` file.

Good point, the code was based on what is done for ocamldep, and should be improved; however the medium term plan is just to don't use `coqdep` but embed a copy of it so we use the ML API and avoid a fork for each file.

File: Makefile.build [hunk 1]
```diff
@@ -897,7 +897,7 @@ endif
 
 $(VDFILE).d: $(D_DEPEND_BEFORE_SRC) $(VFILES) $(D_DEPEND_AFTER_SRC) $(COQDEPBOOT)
 	$(SHOW)'COQDEP    VFILES'
-	$(HIDE)$(COQDEPBOOT) -boot $(DYNDEP) -Q user-contrib "" $(USERCONTRIBINCLUDES) $(VFILES) $(TOTARGET)
+	$(HIDE)$(COQDEPBOOT) -vos -boot $(DYNDEP) -Q user-contrib "" $(USERCONTRIBINCLUDES) $(VFILES) $(TOTARGET)
```
Comment (hendriktews) [inline Makefile.build:900]: AFAICT coqdep_boot does not accept the `-vos` option.

File: tools/CoqMakefile.in [hunk 2]
```diff
@@ -753,7 +753,7 @@ VDFILE_FLAGS:=$(if @PROJECT_FILE@,-f @PROJECT_FILE@,) $(CMDLINE_COQLIBS) $(CMDLI
 
 $(VDFILE): $(VFILES)
 	$(SHOW)'COQDEP VFILES'
-	$(HIDE)$(COQDEP) -dyndep var $(VDFILE_FLAGS) $(redir_if_ok)
+	$(HIDE)$(COQDEP) -vos -dyndep var $(VDFILE_FLAGS) $(redir_if_ok)
```
Comment (hendriktews) [inline tools/CoqMakefile.in:756]: Similarly here.

**Pull request merged**